xunit-autolabeler-v2/ast_parser/python/test_parser.py

```diff
@@ -0,0 +1,225 @@
+# Copyright 2020 Google LLC. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import ast
+import os
+import sys
+from typing import Any, Dict, List, Tuple
+
+
+from ast_parser.core import constants
+from . import drift_test
+
+
+def _get_test_nodes(parsed_nodes: List[Any]) -> List[Any]:
+    """Helper function to extract test methods wrapped in classes
+
+    Args:
+        parsed_nodes (List[ast.AST]): a list of nodes parsed
+                                      from a Python test file
+
+    Returns:
+        List[ast.AST]: a list of possible test nodes
+    """
+
```

Comment:
I prefer having a for loop like this:

```python
ret = []
for node in parsed_nodes:
    if 'ClassDef' in str(type(node)):
        candidate = node.body
    else:
        candidate = node
    if hasattr(candidate, 'name') and candidate.name.startswith('test_'):
        ret.append(candidate)
return ret


```

Reply:
Done.

Reply:
(Edit: had to refactor, since class nodes _and_ their bodies may represent tests. But I kept the for-loop architecture.)

Reply:
Modifying the list in the loop sounds scary.

How about:

```python
candidates = []
for node in parsed_nodes:
    candidates.append(node)
    if 'ClassDef' in str(type(node)):
        # Also add their body
        candidates.append(node.body)
return [node for node in candidates
            if hasattr(node, 'name') and node.name.startswith('test_')]
```